activerecord/lib/active_record/associations/belongs_to_association.rb

```diff
@@ -70,7 +70,10 @@ def replace(record)
 
           replace_keys(record)
 
-          self.target = record
+          if target != record
```

Comment:
The `@inversed` status getting reset after every assignment is what originally lead me to this bug. `Association#target=` resets the inverse status.

When it's the same instance, not touching the inverse status makes sense. Here is a runnable test case to illustrate this problem: https://gist.github.com/XrXr/6cbfbd842e293c4613572c2325320b2d